csharp/src/Microsoft.ML.OnnxRuntime/OrtTypeInfo.shared.cs

```diff
@@ -0,0 +1,342 @@
+        /// Overrides SafeHandle.IsInvalid
+        /// </summary>
+        /// <value>returns true if handle is equal to Zero</value>
+        public override bool IsInvalid { get { return handle == IntPtr.Zero; } }
+
+        /// <summary>
+        /// Overrides SafeHandle.ReleaseHandle() to properly dispose of
+        /// the native instance of OrtValue
+        /// </summary>
+        /// <returns>always returns true</returns>
+        protected override bool ReleaseHandle()
+        {
+            NativeMethods.OrtReleaseTypeInfo(handle);
+            handle = IntPtr.Zero;
+            return true;
+        }
+        #endregion
+    }
+
+    /// <summary>
+    /// This class represents type and shape information for a tensor.
+    /// It may describe a tensor type that is a model input or output or
+    /// an information that can be extracted from a tensor in OrtValue.
+    /// 
+    /// </summary>
+    public class OrtTensorTypeAndShapeInfo : SafeHandle
+    {
+        private readonly bool _owned;
+
+        internal OrtTensorTypeAndShapeInfo(IntPtr handle, bool owned) : base(handle, true)
+        {
+            _owned = owned;
+        }
+
+        internal IntPtr Handle { get { return handle; } }
+
+        /// <summary>
+        /// Fetches tensor element data type
+        /// </summary>
+        /// <value>enum value for the data type</value>
+        public TensorElementType ElementDataType
+        {
+            get
+            {
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorElementType(Handle, out IntPtr type));
+                return (TensorElementType)type;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if this data element is a string
+        /// </summary>
+        public bool IsString
+        {
+            get
+            {
+                return ElementDataType == TensorElementType.String;
+            }
+        }
+
+        /// <summary>
+        /// Fetches tensor element count based on the shape information.
+        /// </summary>
+        /// <returns>number of typed tensor elements</returns>
+        public long GetElementCount()
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorShapeElementCount(Handle, out UIntPtr count));
+            return (long)count;
+        }
+
+        /// <summary>
+        /// Fetches shape dimension count (rank) for the tensor.
+        /// </summary>
+        /// <returns>dim count</returns>
+        public int GetDimensionsCount()
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetDimensionsCount(Handle, out UIntPtr count));
+            return (int)count;
+        }
+
+        /// <summary>
+        /// Fetches tensor symbolic dimensions. Use GetShape to fetch integer dimensions.
+        /// The size of the vector returned is equal to the value returned by GetDimensionsCount.
+        /// Positions that do not have symbolic dimensions will have empty strings.
```

Comment:
Not sure returning an empty string is good enough. It's valid to have a symbolic dimension with no name (neither dim_param or dim_value are set), so you may need to use null for non-symbolic dimensions.

Reply:
Doesn't that equate to positions with empty strings may or may not be symbolic dimensions, and you must also call OrtGetDimensions to know?

And if you don't use a null for dimensions with dim_value doesn't that make ambiguous as to what an empty string means?

Not sure what you mean by 'no null option for C#' though. AFAIK you can insert a null in a string[]. This complies without error `var test = new string[] { "one", null, "three" };`

Reply:
We can certainly do what we want in C# including null.
Providing we solve two issues:
1) C# currently simply broadcasts what C API returns, and this is what it returns. So we need to somehow decide on a C# how to interpret it and insert nulls.
2) Should we choose to interpret something to null, how do we change it in NodeMetdata, it makes use of the same interfaces and it has been our public interface for a while, same as our C API.

Reply:
It would have been better for the C API to return nullptr for symbolic dims so a single call could provide all the relevant information. Unlikely we can change that, so at least the new C# API could do the right thing. Forcing a user to call 2 APIs and combine the results is unnecessarily painful and returning an empty string for both fixed dimensions and 'unknown' symbolic dimensions is misleading.

What's the relationship between OrtTypeInfo and NodeMetadata (which I guess equates to the content of TensorTypeAndShape.SymbolicDimensions when NodeMetadata is for a Tensor)? Will OrtTypeInfo replace NodeMetadata (which is confusingly named anyway given it seems more about model level input/output metadata rather than 'node' metadata)

Reply:
It is hard to replace `NodeMetdata` which has been a public API for a long time.
Both `NodeMetdata` and the type and shape information we obtain from `OrtValue` come through the same C API.
The key difference, `OrtTypeInfo` for `NodeMetadata` is obtained for inputs/outputs and overridable initializers.
For inputs/outputs the information comes from `ValueInfo`. Whereas for `OrtValue` it comes from concrete tensors.
Thus, the latter may not have symbolic dimensions. I think we can simply remove that part from OrtTypeInfo since it is only used for `OrtValue`.